File: codejam/Past Contest/Qualification Round 2010/Rotate/others/rotate.cpp
#include <cstdio>

using namespace std;

int N, K;
char B[52][52];
char T[52][52];
void rotate() {
   for (int r = 0; r < N; ++r)
      for (int c = 0; c < N; ++c)
         T[r][c] = B[N-1-c][r];

   for (int r = 0; r < N; ++r)
      for (int c = 0; c < N; ++c)
         B[r][c] = T[r][c];
}

void fall() {
   for (int c = 0; c < N; ++c) {
      for (int r = N-1; r >= 0; --r) {
         if (B[r][c] != '.') continue;
         int rr;
         for (rr = r; rr >= 0 && B[rr][c] == '.'; --rr);
         if (rr >= 0) {
            B[r][c] = B[rr][c];
            B[rr][c] = '.';
         }
      }
   }
}

void print() {
   for (int r = 0; r < N; ++r) {
      for (int c = 0; c < N; ++c)
         putchar(B[r][c]);
      putchar('\n');
   }
   putchar('\n');
}


int dr[] = {-1, -1, -1,  0,  0, +1, +1, +1};
int dc[] = {-1,  0, +1, -1, +1, -1,  0, +1};

int main(int argc, char* argv[]) {
   int TC;
   scanf("%d", &TC);
   for (int tc = 1; tc <= TC; ++tc) {
      scanf("%d %d", &N, &K);
      for (int i = 0; i < N; ++i)
         scanf("%s", &B[i]);

//      print();

      rotate();

//     print();

      fall();

//      print();

      int winner = 0;
      for (int r = 0; r < N; ++r)
         for (int c = 0; c < N; ++c) {
            char ch = B[r][c];
            if (ch == '.') continue;
            for (int d = 0; d < 8; ++d) {
               int k;
               for (k = 0; k < K; ++k) {
                  int nr = r + dr[d]*k, nc = c + dc[d]*k;
                  if (nr < 0 || nr >= N || nc < 0 || nc >= N)
                     break;
                  if (B[nr][nc] != ch)
                     break;
               }
               if (k >= K)
                  if (ch == 'B') {
                     winner |= 1;
                  //   printf("Blue %d %d  %d %d %d\n", r, c, d, dr[d], dr[c]);
                  }
                  else {
                     winner |= 2;
                  //   printf("Red %d %d  %d %d %d\n", r, c, d, dr[d], dr[c]);
                  }
            }
         }
      printf("Case #%d: ", tc);
      switch (winner) {
         case 0: puts("Neither"); break;
         case 1: puts("Blue"); break;
         case 2: puts("Red"); break;
         default: puts("Both");
      }
   }
   return 0;
}

File: codejam/Past Contest/Qualification Round 2010/Rotate/others/Rotate.java
package google.gcj2010.round1a;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.PrintWriter;

public class Rotate {

    private static boolean DEBUG = false;
    
    private static char[][] B;
   
    private static boolean joink(char P, int K) {
        int c;
        
        // rows
        for (int i = 0; i < B.length; i++) {
            c = 0;
            for (int j = 0; j < B.length; j++) {
                if (B[i][j] == P)
                    c++;
                else
                    c = 0;
                if (c == K)
                    return true;
            }
        }
        // cols
        for (int i = 0; i < B.length; i++) {
            c = 0;
            for (int j = 0; j < B.length; j++) {
                if (B[j][i] == P)
                    c++;
                else
                    c = 0;
                if (c == K)
                    return true;
            }
        }
        //diagonals
        for (int i = 0; i < B.length; i++) {
            c = 0;
            for (int j = 0; j < B.length - i; j++) {
                if (B[i + j][j] == P)
                    c++;
                else
                    c = 0;
                if (c == K)
                    return true;
            }
            c = 0;
            for (int j = 0; j < B.length - i; j++) {
                if (B[i + j][B.length - j - 1] == P)
                    c++;
                else
                    c = 0;
                if (c == K)
                    return true;
            }
            c = 0;
            for (int j = 0; j < B.length - i; j++) {
                if (B[j][i + j] == P)
                    c++;
                else
                    c = 0;
                if (c == K)
                    return true;
            }
            c = 0;
            for (int j = 0; j < B.length - i; j++) {
                if (B[j][B.length - i - 1] == P)
                    c++;
                else
                    c = 0;
                if (c == K)
                    return true;
            }
        }
        return false;
    }

    private static void solve(String file) throws Exception {
        
        BufferedReader f = new BufferedReader(new FileReader(file + ".in"));
        PrintWriter out = new PrintWriter(new BufferedWriter(new FileWriter(file + ".out")));

        int T = Integer.parseInt(f.readLine());

        for (int i = 0; i < T; i++) {
            String[] s = f.readLine().split(" ");
            int N = Integer.parseInt(s[0]);
            int K = Integer.parseInt(s[1]);
            if (DEBUG)
                System.out.println(N + " " + K);
            
            B = new char[N][N];
            for (int j = 0; j < N; j++) {
                String line = f.readLine().replaceAll("\\.", "");
                for (int k = 0; k < N - line.length(); k++)
                    B[j][k] = '.';
                for (int k = 0; k < line.length(); k++)
                    B[j][N - k - 1] = line.charAt(line.length() - k - 1);
            }
            if (DEBUG) {
                for (int j = 0; j < N; j++) {
                    for (int k = 0; k < N; k++)
                        System.out.print(B[j][k]);
                    System.out.println();
                }
            }
            
            String winner = "Neither";
            boolean red = joink('R', K);
            boolean blue = joink('B', K);
            if (red && blue)
                winner = "Both";
            else if (red)
                winner = "Red";
            else if (blue)
                winner = "Blue";

            out.println("Case #" + (i+1) + ": " + winner);
            System.out.println("Case #" + (i+1) + ": " + winner);
        }
        out.close();
    }
    
    public static void main(String[] args) throws Exception {
        
        //String file = "A-test";
        String file = "A-small-attempt0";
        //String file = "A-large";

        solve(file);
    }
}
